### ISSUE TITLE
process startup fails when the backend isn't live at that time

### ISSUE THREAD

**Issue by rbtcollins** [jaegertracing/jaeger#562]
This might be contentious, but I'd prefer to see the process startup be as tolerant of the backend being unavailable as the process is to disconnect + reconnect issues later on.

There's an argument to be made that process startup is special, but it increases the coupling of deployments, and doesn't really save anything, because monitoring has to be catching these issues regardless (e.g. via liveness checks).

Specifically I'm talking about the collector and query components today, where they fail hard if cassandra isn't fully available and configured.

**Comment by yurishkuro** [comment]
> but I'd prefer to see the process startup be as tolerant of the backend being unavailable as the process is to disconnect + reconnect issues later on.

Yes, that's the same reason that Docker docs gives why they don't support health-check based startup order, which is messing up the crossdock tests (#547). We're limited by the behavior of gocql driver here, which cannot init the session until it can connect to the cluster once. We could just loop in main until the first connection succeeds, but we'd need to depend on a specific error returned from fairly deep in the stack to know when we should continue looping and when we should fail hard because it's a configuration error.

**Comment by rbtcollins** [comment]
I've filed a ticket on gocql; we'll see what that team says.

**Comment by jpkrohling** [comment]
> We could just loop in main until the first connection succeeds, but we'd need to depend on a specific error returned from fairly deep in the stack to know when we should continue looping and when we should fail hard because it's a configuration error.

Is this something we'd like to pursue? The same pattern is repeated across all storage mechanisms (kafka,es).

**Comment by yurishkuro** [comment]
Internally we implemented a "lazy session" for Cassandra that defers the actual session creation until a real query is invoked. If creation fails, a dummy session and other entities (like Query) are provided that return errors, but on the top-level query method a new attempt will be made.

@vprithvi were there any issues with this that? If not, we can open source that.

**Comment by jpkrohling** [comment]
I'm starting to play around with this one and I think the lazy loading could be done at the storage factory level, instead of inside individual storage mechanisms. This would also be an opt-in feature, requiring a new flag to be set. How does that sound? This way, we don't break the expectation of current users that are used to getting failures when the backend isn't available.

**Comment by yurishkuro** [comment]
Good idea. Our Cassandra lazy session (I saved it as [gist][]) had to implement larger API than just the span storage. However, in order to do this at the generic factory level, the concrete factories will need to support some form of health check. E.g. in case of Cassandra, succeeding in creating a session is the health check.

[gist]: https://gist.github.com/yurishkuro/d460a21c293cd02306cc8bc19d895940

**Comment by martin-sucha** [comment]
Since I've seen this issue mentioned in https://github.com/gocql/gocql/issues/1027 that I just closed, I'm writing here to let you know that the latest version of gocql (since March, https://github.com/gocql/gocql/pull/1369) has [ReadyPolicy](https://pkg.go.dev/github.com/gocql/gocql#ReadyPolicy) which should enable the use case for startup without reachable CQL servers.
